Alh/locking

wolfsage opened the pull request.

wolfsage committed:
Add test to show concurrent updates can corrupt states

wolfsage committed:
ix_set - lock the state row for the collection being updated

This ensures that updates are atomic for a collection.

We no longer have an overarching transaction around a set of JMAP method
calls.

Instead, we start one transaction inside of ix_set around set/update/destroy
so that a single method call holds a lock and is released when the method
completes.

We also refresh our view of the current account states between method calls
incase they have changed out from under us (or we thought they changed but
dismissed the changes when a method call failed somehow).

wolfsage committed:
Wrap all ix_set/get calls in transactions for consistent data access

In the future we may pick a stricter isolation level on these
transactions.

rjbs commented on lib/Ix/AccountState.pm:
```diff
@@ -83,18 +83,15 @@ sub _save_states ($self) {
   my $pend = $self->_pending_states;
 
+  # XXX - Should only ever be one pending state. Croak if not true? Modify
+  #       code to expect only one? Other? -- alh, 2017-01-27
```
Is this true?  If you update groups, and that causes subscriptions to be updated in a nested transaction, won't there be two pending states?

wolfsage replied:
The nested ix_set would have bumped the state for that row

rjbs replied:
Okay, so I'm looking at this code now, and here's the scenario I imagine:
* ix_set is called for setFoos
* set.create creates some Foos, then calls ensure_state_bumped(foos)
* …so now foos is pending
* set.update updates some Foos
* …so ix_updated is called, altering related Bars with a nested setBars
• …so ix_set on Bars calls  ensure_state_bumped(bars)
* …then when ix_set on Bars is about to return, it calls _save_state, and there are two pending states

In the case I originally thought of, where create A calls create B, we're safe because ix_created (which would do the nested create of B) would be called before A has called ensure_state_bumped.  The scenario I give above gets past that happy coincidence.

wolfsage replied:
This is now resolved with our new plan.

rjbs commented on lib/Ix/AccountState.pm:
```diff
@@ -83,18 +83,15 @@ sub _save_states ($self) {
-      });
-
-      $rows->{$type} = $row;
+      die sprintf "State row for %s %s does not exist?!\n",
```
Rather than doing this, would it be sane to create missing state rows when state object is initialized?  There would be a one-time race, something would win, other requests would be retried, and we'd move on.

wolfsage replied:
Wouldn't it be better to ensure we create new state rows for accounts when we introduce new entities and disallow any bad data from races?

Or do you think the impacts (and likelihood) of such races is minimal?

rjbs replied:
I think both.  We should create those states, but we can also recover if we don't, rather than have an error case that requires manual intervention.

wolfsage replied:
Okay, we can't lock the collection without a state row, but I suspect that just means the call will fail if someone else creates the state row while we're in the middle of trying to

rjbs commented on t/lib/Bakesale.pm:
```diff
@@ -23,13 +23,13 @@ package Bakesale::Test {
     return ($app, $jmap_tester);
   }
 
-  my @TEST_DBS;
```
cute

rjbs commented on t/lib/Bakesale.pm:
```diff
@@ -53,6 +53,8 @@ package Bakesale::Test {
 
     $user1 = $user_rs->single({ id => $user1->id });
 
+    $user1->ix_create_base_state;
```
Do we want to do this implicitly on creates?

wolfsage replied:
Like, creating any row for a class where ix_is_account_base is true automatically calls this? We can but I *think* I tried that and there's not a clean way to do it. When DBIC calls new it passes in a hashref of attrs that we can pull a result source out of, but it feels like we're reaching into DBIC internals.

Up to you

rjbs replied:
Eh, not that invested in it.

wolfsage committed:
Reset signaled before we unblock our children to avoid race

rjbs commented on t/states.t:
```diff
@@ -0,0 +1,255 @@
+use 5.20.0;
```
As discussed, this test file is not passing reliably for me.

wolfsage replied:
Fix pushed, can you see if that's reliable? Thanks

rjbs replied:
Works better now, thanks.

wolfsage committed:
t/states.t - use a unique signal per child to talk to the parent

Identical signals that come in at the same time will be treated
like a single signal, so we must use separate ones to keep
using these as counters.

wolfsage committed:
Only bump state once per collection-type-change, at the end of a method call

If setCakes internally calls setCakeToppers once per cake, at the end we
should only bump cakeTopperState once for the entire collection.

If the internal calls fail (or the cookies themselves fail to make) after
creating the cake toppers, those toppers should be discarded and the
state should not change.

Also wrap ix_gets in these transactions as well - in the future we will
likely set the gets to SERIALIZABLE isolation level so that they get
repeatable reads, but for now this ensures that $ctx->state doesn't
hang around between separate method calls in a JMAP request

rjbs commented on lib/Ix/Context/WithAccount.pm:
```diff
@@ -34,15 +36,92 @@ has root_context => (
+
+  # Start of a tree? Localize state so it goes away when we're done
+  if ($self->_txn_level == 0) {
+    local $self->{state} = $self->_build_state;
```
This `local` seems like an obvious error.  Probably this should be causing a failing test.

wolfsage committed:
Properly localize state so it goes away when txn_do finishes or dies

wolfsage committed:
Not having initial state is no longer a problem

wolfsage committed:
Update test after rebase
